llama: Decouple patching script from submodule (#7139)
ollama/ollama: docs/development.md

```

You will need to ensure your PATH includes go, cmake, gcc and clang mingw32-make to build ollama from source. (typically `C:\msys64\clangarm64\bin\`)

## Vendoring

> Ollama is designed to support multiple LLM backends, and currently utilizes [llama.cpp](https://github.com/ggerganov/llama.cpp/) and [ggml](https://github.com/ggerganov/ggml) through a vendoring model.  While we generally strive to contribute changes back upstream to avoid drift, we cary a small set of patches which are applied to the tracking commit.  A set of make targets are available to aid developers in updating to a newer tracking commit, or to work on changes.

Reviewer: ```suggestion
Ollama currently vendors [llama.cpp](https://github.com/ggerganov/llama.cpp/) and [ggml](https://github.com/ggerganov/ggml) through a vendoring model.  While we generally strive to contribute changes back upstream to avoid drift, we cary a small set of patches which are applied to the tracking commit.  A set of make targets are available to aid developers in updating to a newer tracking commit, or to work on changes.
```

Reviewer 2: I would simplify this even more:

Ollama currently vendors [llama.cpp](https://github.com/ggerganov/llama.cpp/) and [ggml](https://github.com/ggerganov/ggml). We contribute changes back upstream to avoid drift, but there are additional patches which are applied to the tracking commit. A set of make targets are available to update to a newer tracking commit or to work on changes.

> [!IMPORTANT]
> Prior to merging #7157 we continue to leverage a submodule for llama.cpp which establishes the tracking commit.  After merging that PR a new manifest file we be utilized

> If you will be updating the vendoring code, you should start by running the following command to establish the tracking llama.cpp repo at the top of your ollama repo.

Reviewer 2: Maybe:

If you update the vendoring code, start by running the following command to establish the tracking llama.cpp repo at the top of your Ollama repo.

```
make -C llama apply-patches
```

### Updating Base Commit

**Pin to new base commit**

To update to a newer base commit, select the upstream git tag or commit

> [!IMPORTANT]
> After merging #7157 a manifest will be used instead of the submodule

```
cd llm/llama.cpp
git fetch
git checkout NEW_BASE_COMMIT
cd ..
git add llama.cpp
```

> **Apply Ollama Patches**

Reviewer: ```suggestion
## Applying patches
```

When updating to a newer base commit, the existing patches may not apply cleanly and require manual merge resolution.

> Start by applying the patches.  If any of the patches have conflicts, the `git am` will stop at first failure.

Reviewer 2: > stop at first failure

Should be "stop at the first failure"

```
make -C llama apply-patches
```

If you see an error message about a conflict, go into the llama.cpp tracking repo at the top of the ollama repo, and perform merge resolution using your preferred tool to the patch commit which failed.  Save the file(s) and continue the patch series with `git am --continue` .  If any additional patches fail, follow the same pattern until the full patch series is applied.  Once finished, run a final `patch` and `sync` target to ensure everything is updated.

```
make -C llama patch sync
```

Build and test Ollama, and make any necessary changes to the Go code based on the new base commit.  Submit your PR to the Ollama repo.

> ### New Development

Reviewer: ```suggestion
### Generating a new patch
```

Maybe can be it's own section below this, and this could be

```
## Building
```

Author: Went with `Generating Patches`

When working on new fixes or features that impact vendored code, use the following model.  First get a clean tracking repo with all current patches applied:

```
make -C llama apply-patches
```

Now edit the upstream native code in the llama.cpp tracking repo at the top of the ollama repo.  You do not need to commit every change in order to build, a dirty working tree in the tracking repo is OK while developing.  Simply save in your editor, and run the following to refresh the vendored code with your changes, build the backend(s) and build ollama:

```
make -C llama sync
make -C llama -j 8
go build .
```

> [!IMPORTANT]
> Do **NOT** run `apply-patches` while you're iterating as that will reset the tracking repo.  It will detect a dirty tree and abort, but if your tree is clean and you accidentally ran this target, use `git reflog` to recover your commit(s).

Iterate until you're ready to submit PRs.  Once your code is ready, commit a change in the llama.cpp tracking repo, then generate a new patch for ollama with

```

> make -C llama patch

Author: Maybe "gen-patches" or something like that since "patch" is a bit ambiguous?

> make -C llama patch

Reviewer: ```suggestion
make -C llama patches
```

Author: went with `create-patches`
